### ISSUE TITLE
File Comparison

### ISSUE THREAD

**Issue by pchan37** [sbu-ces-unofficial/copy-catch#3]
What are we going to define as a discrepancy?  How should we compare files/directories to find discrepancies without raising false positives while being more thorough than just checking file count and file size?

**Comment by pchan37** [comment]
@morphmeme 
Let's do a few tests to see if the newlines are preserved on copy with the following software:
- [ ] FastCopy
- [ ] Windows Explorer
- [ ] MacOS Drag and Drop
- [ ] Rust lang built-in copy

Also, run a benchmark to see how long it would take to do a diff ignoring newlines on following backup size:
- [ ] 250gb (avg backup size)
- [ ] 500gb (maximum for terrastation)
- [ ] 750gb (a upper limit that should cover most use cases)
- [ ] 1tb (nice to have upper limit)

If line numbers are preserved in all scenarios, we might be able to use a checksum approach.  Otherwise, based on the above benchmark, we have the following options where small file size refers to largest file size we can support w/o stalling progress and large file size refers to anything larger than that:

1) On small file size, do a diff ignoring newlines.  On large file sizes, do a check by file size and file count.  This would still improve integrity check on small file count.
2) Do a diff ignoring newlines regardless of file size.  Improve integrity check always, but could take a large amount of time if we have huge file size.

**Comment by morphmeme** [comment]
I ran benchmarks previously using blake2b hash, and it was pretty slow. Ill run it again with different hashing methods.

Also I doubt that newlines arent copied over, since each copy method probably copies byte by byte, but it should be simple to test anyway.

**Comment by morphmeme** [comment]
Also, the slow down is probably because of reading the entire file. We could probably randomize how much we read in to speed it up with extra security.

**Comment by pchan37** [comment]
> Also I doubt that newlines arent copied over, since each copy method probably copies byte by byte, but it should be simple to test anyway.
Based on prior experience, it's possible for newline characters to be converted from "\n" to "\r\n" which can mess up the checksum calculation.

> Also, the slow down is probably because of reading the entire file. We could probably randomize how much we read in to speed it up with extra security.
I forgot that you need to read in the file to do a checksum calculation.  Guess we can just go with diffing?

**Comment by morphmeme** [comment]
We could calculate the hash in parallel when copying and store the hash to verify. We probably wanna make sure that the difference in file sizes never happens, so that our co workers will not have any concerns about the program. If we are gonna report discrepancies, it should be after the backup.

 I don’t think it is necessary to make an independent discrepancy checker because it would not only use something less secure, its more convenient to solve the whole problem of discrepancies during backup rather than creating another tool to do so.

**Comment by morphmeme** [comment]
Also, since we still need to read the files in the network drive, we can do this in parallel while baxking up

**Comment by pchan37** [comment]
I was thinking of structuring it as a library with an executable bundled in.  Generally speaking in the Unix world, you want each executable to do one thing and do it well.  Also, you limit the flexibility if you just bundle the two tools together.  For example, in CES not every backup is to be done by the backup script.  You need to know how to do it manually and a backup discrepancy tool would still be useful.

> I don’t think it is necessary to make an independent discrepancy checker because it would not only use something less secure, its more convenient to solve the whole problem of discrepancies during backup rather than creating another tool to do so.

If you write the library portion to calculate the hash of two files, then I don't see how that would be less secure than bundling it with the backup script directly.  You can then call that function in the backup script to reuse that portion of the code.

**Comment by morphmeme** [comment]
If the discrepancy reporter is used independently of the backup script, then the hash needs to be calculated which is probably too slowly. We’ll probably have to use diff which is less secure.

I think it’s better that if someone wants the security of a discrepancy checker, then they should have used the backup program instead. Also, we may disagree here, but if we make a discrepancy checker standalone, then people won’t even bother with the program because it’s already pretty easy to fast copy and now since they have a discrepancy checker, they wouldn’t have to bother using our program.

**Comment by pchan37** [comment]
> If the discrepancy reporter is used independently of the backup script, then the hash needs to be calculated which is probably too slowly. We’ll probably have to use diff which is less secure.

How is using the diff less secure?

> I think it’s better that if someone wants the security of a discrepancy checker, then they should have used the backup program instead. Also, we may disagree here, but if we make a discrepancy checker standalone, then people won’t even bother with the program because it’s already pretty easy to fast copy and now since they have a discrepancy checker, they wouldn’t have to bother using our program.

I'm not sure what security you're talking about here.  Also, fast copy only runs on Windows.  We are looking to support MacOSX and Linux.  You seem to forget that you need to do backup manually too.  Are you gonna force people to then run the backup again with our program?  It seems fair to have a backup app that can check for discrepancies using this library as well as have a standalone for when you don't need to do the backup part but need to check for discrepancies.

**Comment by morphmeme** [comment]
I meant diff as in file size/file date/file name or other metadata

**Comment by morphmeme** [comment]
Also drag and drop is pretty easy to do. People are not forced to use our program again. Current backups are not affected because they have already been verified

**Comment by pchan37** [comment]
> Also drag and drop is pretty easy to do. People are not forced to use our program again. Current backups are not affected because they have already been verified

I believe the initial reason for why this was bought up was because it takes a long time to verify deeply nested folders if there's a missing file.  So it would still be helpful in manual backups.  In any case, even if you bundle it with the backup script, what's to force people to use it?  How often is the backup script used these days?

**Comment by morphmeme** [comment]
We can make the problem of having to check deep directories disappear rather than directly solving it by never having discrepancies or reporting it after backup.

The backup script is not used because it’s easier to just fast copy and the backup script fails occasionally. I don’t literally want to force people to use our program. The majority will eventually use our program if it’s more convenient than fastcopy.

**Comment by pchan37** [comment]
Closing this in preference of #4.
